onnxruntime/core/optimizer/transpose_optimizer/transpose_optimizer.cc

```diff
@@ -1751,11 +1848,50 @@ bool OptimizeImpl(OptimizerCtx& ctx) {
       }
     }
   }
+
+  // Run second optimization pass.
+  // If any transpose succeeds a DQ node, move it above the DQ node.
+  // In case of QDQ models this helps to preserve the QDQ node unit
+  // In all other scenarios this is beneficial as well because moving transpose above DQ node is more efficient as
+  // transpose node now handles less data.
+  auto graph_nodes = ctx.graph.Nodes();
+  for (size_t i = 1; i < graph_nodes.size(); i++) {
+    if (graph_nodes[i]->OpType() == "Transpose" && graph_nodes[i - 1]->OpType() == "DequantizeLinear") {
+      auto& transpose_node = *graph_nodes[i];
```

Comment:
I would have expected these to be guaranteed to be adjacent. should we instead check the map to producer node to be sure?